{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_public_HO = pd.read_csv(r'C:\\Users\\91735\\OneDrive\\Desktop\\freelance excel\\Fiverr\\Sachaat\\Files\\v54.1.p1_HO_public.txt', delimiter='\\t',low_memory=False)\n",
    "\n",
    "df_public_HO.to_csv('v54.1.p1_HO_public.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Genetic ID', 'Master ID', 'Skeletal code', 'Skeletal element',\n",
       "       'Year data from this individual was first published [for a present-day individuals we give the data of the data reported here; missing GreenScience 2010 (Vi33.15, Vi33.26), Olalde2018 (I2657), RasmussenNature2010 (Australian)]',\n",
       "       'Publication',\n",
       "       'Method for Determining Date; unless otherwise specified, calibrations use 95.4% intervals from OxCal v4.4.2 Bronk Ramsey (2009); r5; Atmospheric data from Reimer et al (2020)',\n",
       "       'Date mean in BP in years before 1950 CE [OxCal mu for a direct radiocarbon date, and average of range for a contextual date]',\n",
       "       'Date standard deviation in BP [OxCal sigma for a direct radiocarbon date, and standard deviation of the uniform distribution between the two bounds for a contextual date]',\n",
       "       'Full Date One of two formats. (Format 1) 95.4% CI calibrated radiocarbon age (Conventional Radiocarbon Age BP, Lab number) e.g. 2624-2350 calBCE (3990Â±40 BP, Ua-35016). (Format 2) Archaeological context range, e.g. 2500-1700 BCE',\n",
       "       'Age at Death from physical anthropology', 'Group ID', 'Locality',\n",
       "       'Political Entity', 'Lat.', 'Long.', 'Pulldown Strategy', 'Data source',\n",
       "       'No. Libraries',\n",
       "       '1240k coverage (taken from original pulldown where possible)',\n",
       "       'SNPs hit on autosomal targets (Computed using easystats on 1240k snpset)',\n",
       "       'SNPs hit on autosomal targets (Computed using easystats on HO snpset)',\n",
       "       'Molecular Sex', 'Family ID and position within family',\n",
       "       'Y haplogroup (manual curation in terminal mutation format)',\n",
       "       'Y haplogroup (manual curation in ISOGG format)',\n",
       "       'mtDNA coverage (merged data)', 'mtDNA haplogroup if >2x or published',\n",
       "       'mtDNA match to consensus if >2x (merged data)',\n",
       "       'Damage rate in first nucleotide on sequences overlapping 1240k targets (merged data)',\n",
       "       'Sex ratio [Y/(Y+X) counts] (merged data)',\n",
       "       'Library type (minus=no.damage.correction, half=damage.retained.at.last.position, plus=damage.fully.corrected, ds=double.stranded.library.preparation, ss=single.stranded.library.preparation)',\n",
       "       'Libraries', 'ASSESSMENT',\n",
       "       'ASSESSMENT WARNINGS (Xcontam interval is listed if lower bound is >0.005, \"QUESTIONABLE\" if lower bound is 0.01-0.02, \"QUESTIONABLE_CRITICAL\" or \"FAIL\" if lower bound is >0.02) (mtcontam confidence interval is listed if coverage >2 and upper bound is <0.'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_public_HO.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_public_HO_2 = pd.read_csv(r'C:\\Users\\91735\\OneDrive\\Desktop\\freelance excel\\Fiverr\\Sachaat\\Files\\v54.1.p1_HO_public (1).txt', delimiter='\\t',low_memory=False)\n",
    "\n",
    "df_public_HO_2.to_csv('v54.1.p1_HO_public (1).csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3094315     1        0.020130          752566 G A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs7419119     1        0.022518    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs13302957     1        0.024116    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs6696609     1        0.024457    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs8997     1        0.025727    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs9442372     1        0.026288    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs147606383     1        0.026665    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs4970405     1        0.026674    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs11807848     1        0.026711    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs4970421     1        0.028311    ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1320571     1        0.028916    ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  rs3094315     1        0.020130          752566 G A\n",
       "0             rs7419119     1        0.022518    ... \n",
       "1            rs13302957     1        0.024116    ... \n",
       "2             rs6696609     1        0.024457    ... \n",
       "3                rs8997     1        0.025727    ... \n",
       "4             rs9442372     1        0.026288    ... \n",
       "5           rs147606383     1        0.026665    ... \n",
       "6             rs4970405     1        0.026674    ... \n",
       "7            rs11807848     1        0.026711    ... \n",
       "8             rs4970421     1        0.028311    ... \n",
       "9             rs1320571     1        0.028916    ... "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_public_HO_2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  0   1         2             3\n",
      "0         rs3094315   1  0.020130    752566 G A\n",
      "1         rs7419119   1  0.022518    842013 T G\n",
      "2        rs13302957   1  0.024116    891021 G A\n",
      "3         rs6696609   1  0.024457    903426 C T\n",
      "4            rs8997   1  0.025727    949654 A G\n",
      "...             ...  ..       ...           ...\n",
      "597568    rs9786095  24         0  24359931 C T\n",
      "597569   rs67462087  24         0  24394612 A G\n",
      "597570    rs7893048  24         0  24444622 T C\n",
      "597571   rs73628929  24         0  24478403 A G\n",
      "597572  rs111621212  24         0  24505075 G T\n",
      "\n",
      "[597573 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Define the path to your text file\n",
    "file_path = r'C:\\Users\\91735\\OneDrive\\Desktop\\freelance excel\\Fiverr\\Sachaat\\Files\\v54.1.p1_HO_public (1).txt'\n",
    "\n",
    "# Initialize a list to store the rows\n",
    "data_rows = []\n",
    "\n",
    "# Open and read the file line by line\n",
    "with open(file_path, 'r') as file:\n",
    "    for line in file:\n",
    "        # Use regex to split the line where there are 2 or more spaces\n",
    "        row_data = re.split(r'\\s{2,}', line.strip())\n",
    "        \n",
    "        # Append the split data as a new row in the data_rows list\n",
    "        data_rows.append(row_data)\n",
    "\n",
    "# Create a DataFrame from the processed data\n",
    "df_public_HO2 = pd.DataFrame(data_rows)\n",
    "\n",
    "# Display the DataFrame\n",
    "print(df_public_HO2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_public_HO2.head(10)\n",
    "df_public_HO2.to_csv('v54.1.p1_HO_public (2).csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020130</td>\n",
       "      <td>752566 G A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs7419119</td>\n",
       "      <td>1</td>\n",
       "      <td>0.022518</td>\n",
       "      <td>842013 T G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs13302957</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024116</td>\n",
       "      <td>891021 G A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6696609</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024457</td>\n",
       "      <td>903426 C T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs8997</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025727</td>\n",
       "      <td>949654 A G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs9442372</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026288</td>\n",
       "      <td>1018704 A G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs147606383</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026665</td>\n",
       "      <td>1045331 G A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs4970405</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026674</td>\n",
       "      <td>1048955 A G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs11807848</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026711</td>\n",
       "      <td>1061166 T C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs4970421</td>\n",
       "      <td>1</td>\n",
       "      <td>0.028311</td>\n",
       "      <td>1108637 G A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0  1         2            3\n",
       "0    rs3094315  1  0.020130   752566 G A\n",
       "1    rs7419119  1  0.022518   842013 T G\n",
       "2   rs13302957  1  0.024116   891021 G A\n",
       "3    rs6696609  1  0.024457   903426 C T\n",
       "4       rs8997  1  0.025727   949654 A G\n",
       "5    rs9442372  1  0.026288  1018704 A G\n",
       "6  rs147606383  1  0.026665  1045331 G A\n",
       "7    rs4970405  1  0.026674  1048955 A G\n",
       "8   rs11807848  1  0.026711  1061166 T C\n",
       "9    rs4970421  1  0.028311  1108637 G A"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_public_HO2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>GeneticDistance(cM)</th>\n",
       "      <th>BasePair_Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020130</td>\n",
       "      <td>752566 G A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs7419119</td>\n",
       "      <td>1</td>\n",
       "      <td>0.022518</td>\n",
       "      <td>842013 T G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs13302957</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024116</td>\n",
       "      <td>891021 G A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6696609</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024457</td>\n",
       "      <td>903426 C T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs8997</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025727</td>\n",
       "      <td>949654 A G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs9442372</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026288</td>\n",
       "      <td>1018704 A G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs147606383</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026665</td>\n",
       "      <td>1045331 G A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs4970405</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026674</td>\n",
       "      <td>1048955 A G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs11807848</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026711</td>\n",
       "      <td>1061166 T C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs4970421</td>\n",
       "      <td>1</td>\n",
       "      <td>0.028311</td>\n",
       "      <td>1108637 G A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP Chromosome GeneticDistance(cM) BasePair_Position\n",
       "0    rs3094315          1            0.020130        752566 G A\n",
       "1    rs7419119          1            0.022518        842013 T G\n",
       "2   rs13302957          1            0.024116        891021 G A\n",
       "3    rs6696609          1            0.024457        903426 C T\n",
       "4       rs8997          1            0.025727        949654 A G\n",
       "5    rs9442372          1            0.026288       1018704 A G\n",
       "6  rs147606383          1            0.026665       1045331 G A\n",
       "7    rs4970405          1            0.026674       1048955 A G\n",
       "8   rs11807848          1            0.026711       1061166 T C\n",
       "9    rs4970421          1            0.028311       1108637 G A"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_public_HO2.columns = ['SNP', 'Chromosome','GeneticDistance(cM)','BasePair_Position']\n",
    "df_public_HO2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_public_HO2.to_csv('v54.1.p1_HO_public (2).csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed with utf-8: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte\n",
      "Successfully read the file with utf-16 encoding\n"
     ]
    }
   ],
   "source": [
    "encodings = ['utf-8', 'utf-16']\n",
    "\n",
    "for encoding in encodings:\n",
    "    try:\n",
    "        df_chrY_infofile = pd.read_csv(r'C:\\Users\\91735\\OneDrive\\Desktop\\freelance excel\\Fiverr\\Sachaat\\Files\\chrY_infofile_text.txt', delimiter=';',encoding=encoding)\n",
    "        print(f\"Successfully read the file with {encoding} encoding\")\n",
    "        break\n",
    "    except Exception as e:\n",
    "        print(f\"Failed with {encoding}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chrY_infofile.to_csv('chrY_infofile.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genetic ID</th>\n",
       "      <th>Political Entity</th>\n",
       "      <th>Locality</th>\n",
       "      <th>Molecular Sex</th>\n",
       "      <th>Master ID</th>\n",
       "      <th>World Zone</th>\n",
       "      <th>Y haplogroup</th>\n",
       "      <th>mt haplogroup\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ne30_genotyping_noUDG</td>\n",
       "      <td>China</td>\n",
       "      <td>Nepal\\t Kathmandu</td>\n",
       "      <td>M</td>\n",
       "      <td>NE30</td>\n",
       "      <td>Asia (E)</td>\n",
       "      <td>C</td>\n",
       "      <td>D4m\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aconcagua_noUDG.SG</td>\n",
       "      <td>Argentina</td>\n",
       "      <td>Cerro Aconcagua\\t Mendoza Province</td>\n",
       "      <td>M</td>\n",
       "      <td>Aconcagua</td>\n",
       "      <td>America (S)</td>\n",
       "      <td>Q-M3</td>\n",
       "      <td>C1bi \\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I1407</td>\n",
       "      <td>Armenia</td>\n",
       "      <td>Areni 1</td>\n",
       "      <td>M</td>\n",
       "      <td>I1407</td>\n",
       "      <td>Europe (E)</td>\n",
       "      <td>L1a</td>\n",
       "      <td>H\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XAR001</td>\n",
       "      <td>Botswana</td>\n",
       "      <td>Xaro</td>\n",
       "      <td>M</td>\n",
       "      <td>XAR001</td>\n",
       "      <td>Africa (SC)</td>\n",
       "      <td>E1b1a1a1c1a</td>\n",
       "      <td>L3e1a2\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I16444</td>\n",
       "      <td>Channel Islands</td>\n",
       "      <td>Channel Islands\\t Guernsey\\t Vale\\t Le DÃÂ©hus</td>\n",
       "      <td>M</td>\n",
       "      <td>I16444</td>\n",
       "      <td>Europe (W)</td>\n",
       "      <td>I2a1a2</td>\n",
       "      <td>J1c1b1\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NE56_noUDG</td>\n",
       "      <td>China</td>\n",
       "      <td>Amur River Basin</td>\n",
       "      <td>M</td>\n",
       "      <td>NE56</td>\n",
       "      <td>Asia (E)</td>\n",
       "      <td>C2</td>\n",
       "      <td>G2\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CAO011</td>\n",
       "      <td>Cuba</td>\n",
       "      <td>Matanzas City\\t Canimar Abajo</td>\n",
       "      <td>M</td>\n",
       "      <td>CAO011</td>\n",
       "      <td>America (C)</td>\n",
       "      <td>Q1b1a1a1</td>\n",
       "      <td>D1\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CAO014</td>\n",
       "      <td>Cuba</td>\n",
       "      <td>Matanzas City\\t Canimar Abajo</td>\n",
       "      <td>M</td>\n",
       "      <td>CAO014</td>\n",
       "      <td>America (C)</td>\n",
       "      <td>Q1b1a1a1</td>\n",
       "      <td>D1\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CAO017</td>\n",
       "      <td>Cuba</td>\n",
       "      <td>Matanzas City\\t Canimar Abajo</td>\n",
       "      <td>M</td>\n",
       "      <td>CAO017</td>\n",
       "      <td>America (C)</td>\n",
       "      <td>Q1b1a2</td>\n",
       "      <td>..\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CAO022026</td>\n",
       "      <td>Cuba</td>\n",
       "      <td>Matanzas City\\t Canimar Abajo</td>\n",
       "      <td>M</td>\n",
       "      <td>CAO022026</td>\n",
       "      <td>America (C)</td>\n",
       "      <td>Q1b1a1a1</td>\n",
       "      <td>C\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Genetic ID Political Entity  \\\n",
       "0  Ne30_genotyping_noUDG            China   \n",
       "1     Aconcagua_noUDG.SG        Argentina   \n",
       "2                  I1407          Armenia   \n",
       "3                 XAR001         Botswana   \n",
       "4                 I16444  Channel Islands   \n",
       "5             NE56_noUDG            China   \n",
       "6                 CAO011             Cuba   \n",
       "7                 CAO014             Cuba   \n",
       "8                 CAO017             Cuba   \n",
       "9              CAO022026             Cuba   \n",
       "\n",
       "                                        Locality Molecular Sex  Master ID  \\\n",
       "0                              Nepal\\t Kathmandu             M       NE30   \n",
       "1             Cerro Aconcagua\\t Mendoza Province             M  Aconcagua   \n",
       "2                                        Areni 1             M      I1407   \n",
       "3                                           Xaro             M     XAR001   \n",
       "4  Channel Islands\\t Guernsey\\t Vale\\t Le DÃÂ©hus             M     I16444   \n",
       "5                               Amur River Basin             M       NE56   \n",
       "6                  Matanzas City\\t Canimar Abajo             M     CAO011   \n",
       "7                  Matanzas City\\t Canimar Abajo             M     CAO014   \n",
       "8                  Matanzas City\\t Canimar Abajo             M     CAO017   \n",
       "9                  Matanzas City\\t Canimar Abajo             M  CAO022026   \n",
       "\n",
       "    World Zone Y haplogroup  \\\n",
       "0     Asia (E)            C   \n",
       "1  America (S)         Q-M3   \n",
       "2   Europe (E)          L1a   \n",
       "3  Africa (SC)  E1b1a1a1c1a   \n",
       "4   Europe (W)       I2a1a2   \n",
       "5     Asia (E)           C2   \n",
       "6  America (C)     Q1b1a1a1   \n",
       "7  America (C)     Q1b1a1a1   \n",
       "8  America (C)       Q1b1a2   \n",
       "9  America (C)     Q1b1a1a1   \n",
       "\n",
       "  mt haplogroup\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t  \n",
       "0      D4m\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "1    C1bi \\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "2      H\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "3  L3e1a2\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t...         \n",
       "4       J1c1b1\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "5     G2\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "6       D1\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "7       D1\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "8       ..\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         \n",
       "9        C\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t         "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chrY_infofile.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed with utf-8: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte\n",
      "Successfully read the file with utf-16 encoding\n"
     ]
    }
   ],
   "source": [
    "encodings = ['utf-8', 'utf-16']\n",
    "\n",
    "for encoding in encodings:\n",
    "    try:\n",
    "        df_SNP_filtered = pd.read_csv(r'C:\\Users\\91735\\OneDrive\\Desktop\\freelance excel\\Fiverr\\Sachaat\\Files\\chrY_SNP_filtered.txt', delimiter=';',encoding=encoding,low_memory=False)\n",
    "        print(f\"Successfully read the file with {encoding} encoding\")\n",
    "        break\n",
    "    except Exception as e:\n",
    "        print(f\"Failed with {encoding}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_SNP_filtered.to_csv('chrY_SNP_filtered.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed chunk 0 to 10000 with shape (10000, 41006)\n",
      "Processed chunk 10000 to 20000 with shape (10000, 41006)\n",
      "Processed chunk 20000 to 30000 with shape (10000, 41006)\n",
      "Processed chunk 30000 to 40000 with shape (10000, 41006)\n",
      "Processed chunk 40000 to 50000 with shape (10000, 41006)\n",
      "Processed chunk 50000 to 60000 with shape (10000, 41006)\n",
      "Processed chunk 60000 to 70000 with shape (10000, 41006)\n",
      "Processed chunk 70000 to 80000 with shape (10000, 41006)\n",
      "Processed chunk 80000 to 90000 with shape (10000, 41006)\n",
      "Processed chunk 90000 to 100000 with shape (10000, 41006)\n",
      "Processed chunk 100000 to 110000 with shape (10000, 41006)\n",
      "Processed chunk 110000 to 120000 with shape (10000, 41006)\n",
      "Processed chunk 120000 to 130000 with shape (10000, 41006)\n",
      "Processed chunk 130000 to 140000 with shape (10000, 41006)\n",
      "Processed chunk 140000 to 150000 with shape (10000, 41006)\n",
      "Processed chunk 150000 to 160000 with shape (10000, 41006)\n",
      "Processed chunk 160000 to 170000 with shape (10000, 41006)\n",
      "Processed chunk 170000 to 180000 with shape (10000, 41006)\n",
      "Processed chunk 180000 to 190000 with shape (10000, 41006)\n",
      "Processed chunk 190000 to 200000 with shape (10000, 41006)\n",
      "Processed chunk 200000 to 210000 with shape (10000, 41006)\n",
      "Processed chunk 210000 to 220000 with shape (10000, 41006)\n",
      "Processed chunk 220000 to 230000 with shape (10000, 41006)\n",
      "Processed chunk 230000 to 240000 with shape (10000, 41006)\n",
      "Processed chunk 240000 to 250000 with shape (10000, 41006)\n",
      "Processed chunk 250000 to 260000 with shape (10000, 41006)\n",
      "Processed chunk 260000 to 270000 with shape (10000, 41006)\n",
      "Processed chunk 270000 to 280000 with shape (10000, 41006)\n",
      "Processed chunk 280000 to 290000 with shape (10000, 41006)\n",
      "Processed chunk 290000 to 300000 with shape (10000, 41006)\n",
      "Processed chunk 300000 to 310000 with shape (10000, 41006)\n",
      "Processed chunk 310000 to 320000 with shape (10000, 41006)\n",
      "Processed chunk 320000 to 330000 with shape (10000, 41006)\n",
      "Processed chunk 330000 to 340000 with shape (10000, 41006)\n",
      "Processed chunk 340000 to 350000 with shape (10000, 41006)\n",
      "Processed chunk 350000 to 360000 with shape (10000, 41006)\n",
      "Processed chunk 360000 to 370000 with shape (10000, 41006)\n",
      "Processed chunk 370000 to 380000 with shape (10000, 41006)\n",
      "Processed chunk 380000 to 390000 with shape (10000, 41006)\n",
      "Processed chunk 390000 to 400000 with shape (10000, 41006)\n",
      "Processed chunk 400000 to 410000 with shape (10000, 41006)\n",
      "Processed chunk 410000 to 420000 with shape (10000, 41006)\n",
      "Processed chunk 420000 to 430000 with shape (10000, 41006)\n",
      "Processed chunk 430000 to 440000 with shape (10000, 41006)\n",
      "Processed chunk 440000 to 450000 with shape (10000, 41006)\n",
      "Processed chunk 450000 to 460000 with shape (10000, 41006)\n",
      "Processed chunk 460000 to 470000 with shape (10000, 41006)\n",
      "Processed chunk 470000 to 480000 with shape (10000, 41006)\n",
      "Processed chunk 480000 to 490000 with shape (10000, 41006)\n",
      "Processed chunk 490000 to 500000 with shape (10000, 41006)\n",
      "Processed chunk 500000 to 510000 with shape (10000, 41006)\n",
      "Processed chunk 510000 to 520000 with shape (10000, 41006)\n",
      "Processed chunk 520000 to 530000 with shape (10000, 41006)\n",
      "Processed chunk 530000 to 540000 with shape (10000, 41006)\n",
      "Processed chunk 540000 to 550000 with shape (10000, 41006)\n",
      "Processed chunk 550000 to 560000 with shape (10000, 41006)\n",
      "Processed chunk 560000 to 570000 with shape (10000, 41006)\n",
      "Processed chunk 570000 to 580000 with shape (10000, 41006)\n",
      "Processed chunk 580000 to 590000 with shape (10000, 41006)\n",
      "Processed chunk 590000 to 597573 with shape (7573, 41006)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from genetics import loadRawGenoFile, unpackfullgenofile, unpackAndFilterSNPs\n",
    "\n",
    "def process_with_memmap_in_chunks(filename, chunk_size=10000):\n",
    "    geno_file, nind, nsnp, rlen = loadRawGenoFile(filename)\n",
    "    geno = np.memmap(filename, dtype='uint8', mode='r', shape=(nsnp, rlen))\n",
    "    for start in range(0, nsnp, chunk_size):\n",
    "        end = min(start + chunk_size, nsnp)\n",
    "        chunk = geno[start:end]\n",
    "        chunk_unpacked = np.unpackbits(chunk, axis=1)[:, :(2 * nind)]\n",
    "        # Process or save the chunk as needed\n",
    "        # Example: Print the shape of the unpacked chunk\n",
    "        print(f\"Processed chunk {start} to {end} with shape {chunk_unpacked.shape}\")\n",
    "\n",
    "# Example usage\n",
    "process_with_memmap_in_chunks(r'C:\\Users\\91735\\OneDrive\\Desktop\\freelance excel\\Fiverr\\Sachaat\\Files\\v54.1.p1_HO_public.geno')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "freelance",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
